### PR TITLE
Fix crash in ControlFlowGraph related to free functions.

### LINKED ISSUE #11501: [CFG] Throws std::out_of_range exception
```
contract C0 {
  function f() external {}
  function f(int) external {}
  function f2() external {}
}
function f() {}
contract C {
  function f() external {}
}
contract C2 is C0 {}
```

throws the said exception from here

https://github.com/ethereum/solidity/blob/215bbe2e6da7e2b783f04ba10c4d745fbd8c0c7f/libsolidity/analysis/ControlFlowGraph.cpp#L54

Reproduction

```
$ solc test.sol
```

Order of free function declaration matters! If it is declared at the top, exception is not thrown.

### PR DESCRIPTION
**Pull request by Marenz**
fixes #11501

### CODE REVIEW

**Comment by Marenz** [comment]
Test is coming up.

File: Changelog.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ Bugfixes:
  * Code Generator: Fix internal error when super would have to skip an unimplemented function in the virtual resolution order.
  * Control Flow Graph: Take internal calls to functions that always revert into account for reporting unused or unassigned variables.
  * Control Flow Graph: Assume unimplemented modifiers use a placeholder.
+ * Control Flow Graph: Fix crash related to free functions.
```
Comment (chriseth) [inline Changelog.md]: Not needed because the code has not been released yet.

File: libsolidity/analysis/ControlFlowGraph.h [hunk 2]
```diff
@@ -141,9 +141,11 @@ class CFG: private ASTConstVisitor
...
-			if (contract && _other.contract)
-				if (contract->id() != _other.contract->id())
-					return contract->id() < _other.contract->id();
+			int64_t contractId = contract ? contract->id() : -1;
+			int64_t otherContractId = _other.contract ? _other.contract->id() : -1;
+
+			if (contractId != otherContractId)
+				return contractId < otherContractId;
 
 			return function->id() < _other.function->id();
```
Comment (chriseth) [inline libsolidity/analysis/ControlFlowGraph.h]:
```
return
    make_pair(contract ? -1 : contract->id(), function->id()) <
    make_pair(_other.contract ? -1 : _other.contract->id(), _other.function->id());

Comment (chriseth) [reply libsolidity/analysis/ControlFlowGraph.h]: I'm actually wondering if there is a tool somewhere that generates comparison operators based on projections, i.e. something like
```
return KeyComparator{[](FunctionContractTuple const& _x) { return {_x ? -1 : _x->id(), _x.function->id()}}(*this, _other);
```

File: libsolidity/analysis/ControlFlowGraph.h [hunk 3]
```diff
@@ -141,11 +141,9 @@ class CFG: private ASTConstVisitor
...
 		bool operator<(FunctionContractTuple const& _other) const
 		{
-			if (contract && _other.contract)
-				if (contract->id() != _other.contract->id())
-					return contract->id() < _other.contract->id();
-
-			return function->id() < _other.function->id();
+			return
+				std::make_pair(contract ? -1 : contract->id(), function->id()) <
+				std::make_pair(_other.contract ? -1 : _other.contract->id(), _other.function->id());
```
Comment (chriseth) [inline libsolidity/analysis/ControlFlowGraph.h]:
```suggestion
				std::make_pair(contract ? contract->id() : -1, function->id()) <
				std::make_pair(_other.contract ? _other.contract->id() : -1, _other.function->id());
```

**Review by chriseth — APPROVED** [review]

**Pull request merged**